src/main/java/htsjdk/samtools/reference/AbstractFastaSequenceFile.java

```diff
@@ -60,17 +60,7 @@
     AbstractFastaSequenceFile(final Path path) {
         this.path = path;
         this.source = path == null ? "unknown" : path.toAbsolutePath().toString();
-        final Path dictionary = findSequenceDictionary(path);
-
-        if (dictionary != null) {
-            IOUtil.assertFileIsReadable(dictionary);
-            try (InputStream dictionaryIn = Files.newInputStream(dictionary)) {
-                this.sequenceDictionary = ReferenceSequenceFileFactory.loadDictionary(dictionaryIn);
-            }
-            catch (Exception e) {
-                throw new SAMException("Could not open sequence dictionary file: " + dictionary, e);
-            }
-        }
+        this.sequenceDictionary = findAndLoadSequenceDictionary(path);
```

Comment:
Putting a method call to an overridable method in a constructor means that you can end up calling a method in an object whose constructor hasn't been called yet. This can lead to some very hard to debug errors.

One way to avoid this kind of dependency is to make a field lazily initialized, e.g. in its getter.

Reply:
My first thought was indeed to make this lazy loading, but that's not possible without a breaking change.  Unfortunately we have:

```
    protected SAMSequenceDictionary sequenceDictionary;
```

and subclasses (including any outside htsjdk) can access via property access.  If others (@yfarjoun , @lbergelson) are happy to have a breaking change, I'd be happy to rework this PR to make that a `private Lazy<SAMSequenceDictionary>` and move access to be via a method.  Otherwise I'm sorta stuck doing it in the manner I have.

Reply:
sounds good to me.

Reply:
Ok, I've pushed up a change that makes it lazy.  Fixed any problems in htsjdk, and double checked that there aren't classes in Picard or GATK that access that property directly.  @lbergelson are you on board with this?